src/box/lua/config/configdata.lua

```diff
@@ -201,6 +203,146 @@ local mt = {
+       config_names.replicaset_uuid ~= snap_names.replicaset_uuid then
+        error(string.format('Replicaset UUID mismatch. Snapshot: %s, ' ..
+                            'config: %s.', snap_names.replicaset_uuid,
+                            config_names.replicaset_uuid), 0)
+    end
+
+    if snap_names.replicaset_name ~= nil and
+       snap_names.replicaset_name ~= config_names.replicaset_name then
+        error(string.format('Replicaset name mismatch. Snapshot: %s, ' ..
+                            'config: %s.', snap_names.replicaset_name,
+                            config_names.replicaset_name), 0)
+    end
+
+    if config_names.instance_uuid ~= nil and
+       config_names.instance_uuid ~= snap_names.instance_uuid then
+        error(string.format('Instance UUID mismatch. Snapshot: %s, ' ..
+                            'config: %s.', snap_names.instance_uuid,
+                            config_names.instance_uuid), 0)
+    end
+
+    if snap_names.instance_name ~= nil and
+       snap_names.instance_name ~= config_names.instance_name then
+        error(string.format('Instance name mismatch. Snapshot: %s, ' ..
+                            'config: %s.', snap_names.instance_name,
+                            config_names.instance_name), 0)
+    end
+
+    -- Fail early, if current UUID is not set, but no name is found
+    -- inside the snapshot file.
+    --
+    -- TODO: it's NoOp for now, as instance_uuid is automatically
+    -- generated for vshard configuration. This error and error_uuids
+    -- can't happen and consequently not tested. We must get rid of
+    -- generating UUIDs and test it properly. Anonymous replicas should
+    -- also be tested.
+    --
+    -- For now it's impossible to recover from snaps without manually
+    -- specifying UUIDs.
+    local iconfig = config_names.peers[config_names.instance_name].iconfig_def
+    if snap_names.instance_name == nil and
+       config_names.instance_uuid == nil and
+       not instance_config:get(iconfig, 'replication.anon') then
+        error(string.format('Instance name is not set in snapshot and UUID ' ..
+                            'is missing in the config. Found %s in snapshot.',
+                            snap_names.instance_uuid), 0)
+    end
+
+    -- Check, that every peer, which doesn't have name in snap,
+    -- has UUID in config, so that we can validate, that the right
+    -- config is applied to the instance.
+    local warn_uuids = {}
+    local missing_names = {}
+    for _, snap_peer in ipairs(snap_names.peers) do
+        if snap_peer.instance_name ~= nil then
+            goto skip
+        end
+
+        assert(snap_peer.instance_uuid ~= nil)
+        local config_peer_name = find_peer_name_by_uuid(config_names.peers,
+                                                        snap_peer.instance_uuid)
+        if config_peer_name ~= nil then
+            missing_names[config_peer_name] = snap_peer.instance_uuid
```

Comment:
Nit: This is not validation, but rather a part of the logic around `missing_names` consistency. I think that it worth to extract it from this function.

Reply:
I mean, add some separate `configdata:_find_missing_instance_names()`.

Maybe it also worth to add some `configdata:on_instance_name_found()` for better incapsulation.

Reply:
It seems, this comment is now actual anymore.